pkg/controller/awsfederatedrole/awsfederatedrole_controller.go

```diff
@@ -0,0 +1,269 @@
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileAWSFederatedRole{client: mgr.GetClient(),
+		scheme:           mgr.GetScheme(),
+		awsClientBuilder: awsclient.NewClient}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("awsfederatedrole-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource AWSFederatedRole
+	err = c.Watch(&source.Kind{Type: &awsv1alpha1.AWSFederatedRole{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileAWSFederatedRole implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileAWSFederatedRole{}
+
+// ReconcileAWSFederatedRole reconciles a AWSFederatedRole object
+type ReconcileAWSFederatedRole struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client           client.Client
+	scheme           *runtime.Scheme
+	awsClientBuilder func(awsAccessID, awsAccessSecret, token, region string) (awsclient.Client, error)
+}
+
+// Reconcile reads that state of the cluster for a AWSFederatedRole object and makes changes based on the state read
+// and what is in the AWSFederatedRole.Spec
+func (r *ReconcileAWSFederatedRole) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling AWSFederatedRole")
+
+	// Fetch the AWSFederatedRole instance
+	instance := &awsv1alpha1.AWSFederatedRole{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// If the CR is known to be Valid or Invalid, doesn't need to be reconciled.
+	if instance.Status.State == awsv1alpha1.AWSFederatedRoleStateValid || instance.Status.State == awsv1alpha1.AWSFederatedRoleStateInvalid {
+		return reconcile.Result{}, nil
+	}
+
+	// Setup AWS client
+	awsClient, err := awsclient.GetAWSClient(r.client, awsclient.NewAwsClientInput{
+		SecretName: awsSecretName,
+		NameSpace:  awsv1alpha1.AccountCrNamespace,
+		AwsRegion:  "us-east-1",
+	})
+
+	// Validates Custom IAM Policy
+
+	// Build custom policy in AWS-valid JSON and converts to string
+	jsonPolicy, err := utils.YamlToJSONIAMPolicy(*instance)
```

Comment:
why *AWSFederatedRole instead of just passing in AWSFederatedRole?

Reply:
Instance itself is a *AWSFederatedRole, so this is dereferencing it. So this is actually passing AWSFederatedRole :)